### PR TITLE
#5686 Fix issue "Drush installation is not detected for other site al…

### PR DESCRIPTION
**Pull request by ordermind**
…ias when having multiple local Drupal installations with aliases"

### CODE REVIEW

**Comment by ordermind** [comment]
fixes #5686

File: src/Runtime/Runtime.php [hunk 1]
```diff
@@ -55,11 +55,11 @@ public function run($argv): int
     protected function doRun($argv, $output): int
     {
         // Do the preflight steps
-        $status = $this->preflight->preflight($argv);
+        $preflightSuccess = $this->preflight->preflight($argv);
```
Comment (greg-1-anderson) [inline src/Runtime/Runtime.php:58]: This code should remain as it was.

File: src/Preflight/RedispatchToSiteLocal.php [hunk 2]
```diff
@@ -58,7 +58,12 @@ function ($item) {
         );
         $command .= ' ' . implode(' ', $args);
         passthru($command, $status);
-        return $status;
+
+        // The $status variable is the exit status of the process, which is an
+        // integer. This method is supposed to return true when successful, so
+        // we need to convert it to a boolean accordingly.
+
+        return $status ? false : true;
```
Comment (greg-1-anderson) [inline src/Preflight/RedispatchToSiteLocal.php:66]: This code should remain as it was.

File: src/Runtime/Runtime.php (hunk 1 above)
Comment (greg-1-anderson) [inline src/Runtime/Runtime.php]: `false` does not indicate failure, so I would rename this something like `$preflightDidRedispatch`

File: src/Runtime/Runtime.php [hunk 3]
```diff
@@ -55,11 +55,11 @@ public function run($argv): int
...
 
         // If preflight signals that we are done, then exit early.
-        if ($status) {
-            return $status;
+        if ($preflightSuccess) {
```
Comment (greg-1-anderson) [inline src/Runtime/Runtime.php]: Here is the real bug. The original code was [if ($status !== false)](https://github.com/drush-ops/drush/commit/1a7704edfe99a7a223f65154a823318e2ca09499#diff-fcb54a74a331c6112bd5cf5e32f08a4d0a39f42c16833b899b4fac0339d453c7R219). That is, Preflight would return `bool false` if `doRun` should continue, and an integer status code if a redispatch happened. This sort of coding convention was common in the PHP 5 era, but is no longer considered de rigor in PHP 8, and in fact is a deprecated practice.

File: src/Preflight/RedispatchToSiteLocal.php (hunk 2 above)
Comment (greg-1-anderson) [inline src/Preflight/RedispatchToSiteLocal.php]: This function should always return true if it gets to this point. The comment in the header that says "and was returned successfully" was in error. See my other comment for a description of how the code used to work.

I think that the best fix would be to change this function to use `return [false, 0];` wherever it currently uses `return false`, and change this line to `return [true, $status];`. Then, `doRun` must record both return values, call them `$preflightDidRedispatch` and `$status` or something similar. It should then return `$status` if `$preflightDidRedispatch` is true.

**Review by greg-1-anderson — CHANGES_REQUESTED** [review]
Thanks for looking at this. The original bug was caused by optimizing away a type check (`!== false`). I suggest a better fix would be to change the return tupe of the preflight function. Since preflight is an internal function, I don't think we need to consider this a breaking change.

File: src/Runtime/Runtime.php (hunk 1 above)
Comment (ordermind) [reply src/Runtime/Runtime.php]: Thanks for the feedback, this should be fixed now.

File: src/Runtime/Runtime.php (hunk 3 above)
Comment (ordermind) [reply src/Runtime/Runtime.php]: Thanks for the feedback, this should be fixed now.

File: src/Preflight/RedispatchToSiteLocal.php (hunk 2 above)
Comment (ordermind) [reply src/Preflight/RedispatchToSiteLocal.php]: Thanks for the feedback, this should be fixed now.

**Review by greg-1-anderson — APPROVED** [review]
LGTM, Thank you for working on this problem.

@weitzman Does this resolution look OK to you?

**Comment by weitzman** [comment]
Yes, LGTM. The one suggesion I have is to replace the various new `0` with `\Drush\Commands\DrushCommands::EXIT_SUCCESS`

**Review by greg-1-anderson — CHANGES_REQUESTED** [review]
Use \Drush\Commands\DrushCommands::EXIT_SUCCESS as suggested by @weitzman

**Review by weitzman — APPROVED** [review]

**Pull request merged**